### PR TITLE
NEX-155: ddev support

### PR DESCRIPTION
**Pull request by vermario**
This PR adds support for ddev for the starterkit.

### Changes introduced

* There are now two setup scripts, one for lando, one for ddev
* I replicated our custom lando commands (to install a recipe, to set up oauth keys) for ddev

For the ddev setup, I used this article as a guide, and I went with the first option described there: https://www.velir.com/ideas/2024/05/13/how-to-run-headless-drupal-and-nextjs-on-ddev?utm_source=linkedin&utm_medium=social&sc_camp=FFCB08E97FEE4B629943BBC8546739F4

So we are not setting up a different container, but using the web one on a different port. Also, I didn't set up a daemon like they suggest in the article, because I would like this to work like it does in lando, where to see the site you have to use either `ddev npm run dev` or `ddev npm run start`, since this is used to test different things.

One difference is that for the `ddev npm` command to work, you need to cd to the `next` folder first. I tried to highlight this in the README file, but I can see this confusing people.

### How to test

1. [install ddev if you don't have it](https://ddev.readthedocs.io/en/stable/users/install/ddev-installation/)
2. run `./setup-ddev.sh`
3. open backend, see that preview works
4. open frontend (see provided url, note that localhost:3000 does not work
5. open elasticvue, add predefined cluster check that you can see elasticsearch contents
6. `ddev describe` should say this: 
```
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Project: next-drupal-starterkit ~/www/next-drupal-starterkit https://next-drupal-starterkit.ddev.site                                                                         │
│ Docker platform: docker-desktop                                                                                                                                               │
│ Router: traefik                                                                                                                                                               │
├────────────┬──────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬──────────────────────┤
│ SERVICE    │ STAT │ URL/PORT                                                                                                                           │ INFO                 │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ web        │ OK   │ https://next-drupal-starterkit.ddev.site                                                                                           │ drupal PHP8.2        │
│            │      │ InDocker: web:3000,443,80,8025                                                                                                     │ nginx-fpm            │
│            │      │ Host: 127.0.0.1:61141,61140                                                                                                        │ docroot:'drupal/web' │
│            │      │                                                                                                                                    │ Perf mode: mutagen   │
│            │      │                                                                                                                                    │ NodeJS:20            │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ db         │ OK   │ InDocker: db:3306                                                                                                                  │ mariadb:10.11        │
│            │      │ Host: 127.0.0.1:61139                                                                                                              │ User/Pass: 'db/db'   │
│            │      │                                                                                                                                    │ or 'root/root'       │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ elasticsea │ OK   │ https://next-drupal-starterkit.ddev.site:9201                                                                                      │                      │
│ rch        │      │ InDocker: elasticsearch:9200,9200,9300,9300                                                                                        │                      │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ elasticvue │ OK   │ http://next-drupal-starterkit.ddev.site:9003                                                                                       │                      │
│            │      │ InDocker: elasticvue:80                                                                                                            │                      │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ Mailpit    │      │ Mailpit: https://next-drupal-starterkit.ddev.site:8026                                                                             │                      │
│            │      │ `ddev mailpit`                                                                                                                     │                      │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ Next.js    │      │ InDocker: localhost:3000 https://next-drupal-starterkit.ddev.site:3000 http://next-drupal-starterkit.ddev.site:2999                │                      │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ All URLs   │      │ https://next-drupal-starterkit.ddev.site, https://127.0.0.1:61140, http://next-drupal-starterkit.ddev.site, http://127.0.0.1:61141 │                      │
└────────────┴──────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴──────────────────────┘

 ```

### Open questions

Something that is problematic I think, is that `npm run dev` will print localhost:3000 as the url that the frontend runs in, but that path is not accessible. you have to use `https://next-drupal-starterkit.ddev.site:3000` instead, and that information is not so easily accessible...

I wonder if we can improve this somehow?

### CODE REVIEW

File: .ddev/config.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  - name: "Next.js"
+    container_port: 3000
+    http_port: 2999
+    https_port: 3000
+
+hooks:
+  post-start:
+    - exec: "elasticsearch-plugin install analysis-icu"
+      service: elasticsearch
+    - exec: "cd next && npm i"
```
Comment (vermario) [inline .ddev/config.yaml:53]: This gets executed each time you start the environment so having `npm ci` would have taken forever.

File: .ddev/docker-compose.elasticsearch_extras.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#ddev-generated
+services:
+  elasticsearch:
+    environment:
+      http.cors.enabled: "true"
+      http.cors.allow-origin: "'*'"
+      # Disable authentication in local development
+      xpack.security.enabled: "false"
+      xpack.security.enrollment.enabled: "false"
```
Comment (vermario) [inline .ddev/docker-compose.elasticsearch_extras.yaml:9]: This is needed for elasticvue to be able to connect, avoiding CORS issues

File: .ddev/docker-compose.elasticvue.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+services:
+  elasticvue:
+    image: "cars10/elasticvue:1.0.4"
+    #command: /docker-entrypoint.sh nginx -g 'daemon off;'
+    container_name: ddev-${DDEV_SITENAME}-elasticvue
+    labels:
+      com.ddev.site-name: ${DDEV_SITENAME}
+      com.ddev.approot: ${DDEV_APPROOT}
+    environment:
+     - 'ELASTICVUE_CLUSTERS=[{"name": "dev cluster", "uri": "http://${DDEV_SITENAME}.ddev.site:9200/", "password": "", "user": ""}]'
```
Comment (vermario) [inline .ddev/docker-compose.elasticvue.yaml:10]: This way the elasticsearch url will be available as a predefined cluster:

image

Handy!

File: README.md
Comment (vermario) [inline README.md:1]: (this is better reviewed using the "view rich diff" button here: 
image

File: drupal/web/sites/default/settings.php [hunk 4]
```diff
@@ -8,16 +8,20 @@
...
+    'database' => $_ENV['DB_NAME_DRUPAL'],
+    'username' => $_ENV['DB_USER_DRUPAL'],
+    'password' => $_ENV['DB_PASS_DRUPAL'],
+    'prefix' => '',
+    'host' => $_ENV['DB_HOST_DRUPAL'],
+    'port' => '3306',
+    'namespace' => 'Drupal\\Core\\Database\\Driver\\mysql',
+    'driver' => 'mysql',
+  ];
+}
```
Comment (vermario) [inline drupal/web/sites/default/settings.php:24]: ddev has its own way of setting up the db.

File: drupal/web/sites/default/settings.php [hunk 5]
```diff
@@ -69,6 +73,7 @@
 
   case 'local':
   case 'lando':
+  case 'ddev':
```
Comment (vermario) [inline drupal/web/sites/default/settings.php:76]: this works because I have added that same env var to the ddev setup.

File: setup-ddev.sh [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  done
+
+  # All commands were successful. Remove the status file, show messages to the user, and start the frontend site.
+  rm -f "$status_file"
+  echo '🚀 All done!'
+  ddev describe
+  echo '↪️ Use this link to log into the backend as user 1:'
+  ddev drush uli
+  echo '🏎️ Starting the frontend site in production mode...'
+  echo '⚠️️️️⚠️️️️⚠️️️️ Note: the site will be available at https://next-drupal-starterkit.ddev.site:3000, not localhost:3000 as the output of next start suggests.'
```
Comment (vermario) [inline setup-ddev.sh:75]: This is my attempt of getting the attention to the fact that localhost:3000 does not work. maybe there's some better way?

**Review by hkirsman — APPROVED** [review]

**Pull request merged**